[Dynamo] Enable torch function dispatch on HOPs

mlazos opened the pull request:
Enables `__torch_function__` mode dispatch on higher-order operators. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

zou3519 commented on test/dynamo/test_modes.py:
```diff
@@ -660,6 +660,43 @@ def forward(self, x):
+                    kwargs = {}
+
+                if func == torch._higher_order_ops.flex_attention:
+                    raise RuntimeError("test")
```
nit: it's a bit better to subclass RuntimeError and then check that we are raising that subclass

mlazos commented on test/dynamo/test_modes.py:
```diff
@@ -660,6 +660,43 @@ def forward(self, x):
+        import torch._higher_order_ops
+        from torch.nn.attention.flex_attention import flex_attention
+
+        torch.set_default_device(device_type)
```
Let me make this a context manager, oops.

ydwu4 commented on test/dynamo/test_base_hop.py:
```diff
@@ -58,9 +58,9 @@ def f(x, y):
 class GraphModule(torch.nn.Module):
-    def forward(self, L_x_: "f32[3, 3]", L_y_: "f32[3, 3]"):
-        l_x_ = L_x_
+    def forward(self, L_y_: "f32[3, 3]", L_x_: "f32[3, 3]"):
```
It would be better if we could keep the ordering unchanged. It's already quite cognitive load bearing to chase the proxy names to debug the hop subgraphs.

zou3519 commented on test/dynamo/test_base_hop.py:
```diff
@@ -209,9 +209,9 @@ def f(x, y):
 class GraphModule(torch.nn.Module):
-    def forward(self, L_x_: "f32[3, 3]", L_y_: "f32[3, 3]"):
-        l_x_ = L_x_
+    def forward(self, L_y_: "f32[3, 3]", L_x_: "f32[3, 3]"):
```
Inputs got reversed here, is that still expected?

mlazos replied:
Found the issue pushing a PR before this one in a new stack

mlazos committed:
[Dynamo] Fix arg ordering in tf modes

ghstack-source-id: 91408154cbe44f770afeeec102692651ad5f4a77
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159707

mlazos committed:
[Dynamo] Enable torch function dispatch on HOPs

ghstack-source-id: 391eb352a7ac7cfec1a17ae0bcbdab7cc0b4d0cb
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159708
